source/dem/pp_linear_force.cc

```diff
@@ -2,100 +2,166 @@
+                 j))));
+
+        this->effective_shear_modulus[i][j] =
+          (dem_parameters.physical_properties.youngs_modulus_particle.at(i) *
+           dem_parameters.physical_properties.youngs_modulus_particle.at(j)) /
+          (2 *
+           ((dem_parameters.physical_properties.youngs_modulus_particle.at(j) *
+             (2 -
+              dem_parameters.physical_properties.poisson_ratio_particle.at(i)) *
+             (1 + dem_parameters.physical_properties.poisson_ratio_particle.at(
+                    i))) +
+            (dem_parameters.physical_properties.youngs_modulus_particle.at(i) *
+             (2 -
+              dem_parameters.physical_properties.poisson_ratio_particle.at(j)) *
+             (1 + dem_parameters.physical_properties.poisson_ratio_particle.at(
+                    j)))));
+
+        this->effective_coefficient_of_restitution[i][j] =
+          2 *
+          dem_parameters.physical_properties.restitution_coefficient_particle
+            .at(i) *
+          dem_parameters.physical_properties.restitution_coefficient_particle
+            .at(j) /
+          (dem_parameters.physical_properties.restitution_coefficient_particle
+             .at(i) +
+           dem_parameters.physical_properties.restitution_coefficient_particle
+             .at(j));
+
+        this->effective_coefficient_of_friction[i][j] =
+          2 *
+          dem_parameters.physical_properties.friction_coefficient_particle.at(
+            i) *
+          dem_parameters.physical_properties.friction_coefficient_particle.at(
+            j) /
+          (dem_parameters.physical_properties.friction_coefficient_particle.at(
+             i) +
+           dem_parameters.physical_properties.friction_coefficient_particle.at(
+             j));
+
+        this->effective_coefficient_of_rolling_friction[i][j] =
+          2 *
+          dem_parameters.physical_properties
+            .rolling_friction_coefficient_particle.at(i) *
+          dem_parameters.physical_properties
+            .rolling_friction_coefficient_particle.at(j) /
+          (dem_parameters.physical_properties
+             .rolling_friction_coefficient_particle.at(i) +
+           dem_parameters.physical_properties
+             .rolling_friction_coefficient_particle.at(j));
+      }
+}
+
 template <int dim>
 void
 PPLinearForce<dim>::calculate_pp_contact_force(
   std::unordered_map<int, std::unordered_map<int, pp_contact_info_struct<dim>>>
-    *local_adjacent_particles,
+    &local_adjacent_particles,
   std::unordered_map<int, std::unordered_map<int, pp_contact_info_struct<dim>>>
-    *                                               ghost_adjacent_particles,
-  const Parameters::Lagrangian::PhysicalProperties &physical_properties,
-  const double &                                    dt)
+    &           ghost_adjacent_particles,
+  const double &dt)
 {
   // Updating contact force of particles in local-local and local-ghost contact
```

Comment:
Didn't understand this sentence. Shouldn't it be "particles if local-local ..." instead of" in local-local"

Reply:
There are two types of particle-particle contacts: local-local and local-ghost. I changed the sentence to "for local-local and local-ghost ..."